### PR TITLE
Add helm chart for Kubernetes installation

### PR DESCRIPTION
**Pull request by mrtndwrd**
This is a re-implementation of https://github.com/zulip/docker-zulip/pull/287 where the helm chart is a bit more flexible.

1. It uses dependency charts from the Bitnami repository. These charts allow the user to configure a lot more than the first implementation did.
2. It uses a `StatefulSet` because it includes a persistent volume
3. I updated anything that didn't work anymore (like the command to create a realm). 
4. I added the possibility to run `postSetup` scripts which are mounted to `/data/post-setup.d`
5. I added `SOCIAL_AUTH_OIDC_ENABLED_IDPS` to the entrypoint script because I needed it to configure OIDC. Basically this is a bugfix that does not relate to installing with a helm chart.

### CODE REVIEW

**Comment by varac** [comment]
Hi, I'm a colleague of @mrtndwrd and reviewed this so far. 
Following the README.md instructions for this chart, everything installed fine. I just couldn't create a realm, will comment on the specific code now.

File: kubernetes/chart/zulip/templates/NOTES.txt
Comment (varac) [inline kubernetes/chart/zulip/templates/NOTES.txt:14]: I can't make this work for me. I followed the README.md and all pods install fine, I can reach the login page.
But I can't create a realm: 

```
2. To create a realm so you can sign in:

  kubectl -n zulip exec -it $(kubectl get pods --namespace zulip -l "app.kubernetes.io/instance=zulip-zulip" -o jsonpath="{.items[0].metadata.name}") -c zulip -- sudo -u zulip /home/zulip/deployments/current/manage.py generate_realm_creation_link
```

The instance label filter needs to change from ` "app.kubernetes.io/instance=zulip-zulip"` to ` "app.kubernetes.io/instance=zulip"` otherwise it won't match any pod.
Even with the changed filter, `"{.items[0].metadata.name}"` will give me the memcached pod, not the main zulip one:

```
$ kubectl get pods --namespace zulip -l "app.kubernetes.io/instance=zulip" -o jsonpath="{.items[0].metadata.name}" 
zulip-memcached-84ffbdf7d6-xr4nl
```

which will lead to this:

```
$ kubectl -n zulip exec -it $(kubectl get pods --namespace zulip -l "app.kubernetes.io/instance=zulip" -o jsonpath="{.items[0].metadata.name}") -c zulip -- sudo -u zulip /home/zulip/deployments/current/manage.py generate_realm_creation_link 
Error from server (BadRequest): container zulip is not valid for pod zulip-memcached-84ffbdf7d6-xr4nl
```

**Review by varac — CHANGES_REQUESTED** [review]

File: kubernetes/chart/zulip/templates/NOTES.txt
Comment (varac) [reply kubernetes/chart/zulip/templates/NOTES.txt:14]: Using `"app.kubernetes.io/name=zulip"` works: 

```
$ kubectl -n zulip exec -it $(kubectl get pods --namespace zulip -l "app.kubernetes.io/name=zulip" -o jsonpath="{.items[0].metadata.name}") -c zulip -- sudo -u zulip /home/zulip/deployments/current/manage.py generate_realm_creation_link
Please visit the following secure single-use link to register your 
new Zulip organization:

    https://zulip.varac-test.stackspin.net/new/...
```

Comment (mrtndwrd) [reply kubernetes/chart/zulip/templates/NOTES.txt:14]: Thanks for the suggestion, fixed.

**Comment by varac** [comment]
Nice, just tested from scratch again and it works for me now flawlessly. Looking forward to have this merged!

**Comment by mrtndwrd** [comment]
For anyone that stumbles upon this PR, we've had some review-related discussion in Zulip [here](https://chat.zulip.org/#narrow/stream/21-provision-help/topic/K8.20and.20Helm/near/1285089). I'm now waiting for comments on the content of this PR.

**Comment by mrtndwrd** [comment]
I rebased against `main`. I also added a whole bunch of comments to `values.yaml` as requested in the [Zulip channel](https://chat.zulip.org/#narrow/stream/21-provision-help/topic/K8.20and.20Helm).

**Comment by varac** [comment]
@tg44 or anybody else from the zulip team: Any chance to get this merged soon ?

File: kubernetes/chart/zulip/.gitignore
Comment (scratchmex) [inline kubernetes/chart/zulip/.gitignore:2]: any reason for this? I think it is expected to commit this. See https://github.com/prometheus-community/helm-charts/tree/main/charts/kube-prometheus-stack for reference

File: kubernetes/chart/zulip/values.yaml
Comment (scratchmex) [inline kubernetes/chart/zulip/values.yaml:126]: is the docker image currently made multiplatform? e.g. arm? if not maybe we should consider adding a default affinity for those architectures

**Review by scratchmex — COMMENTED** [review]
I think this is a very good Helm chart. I just made some comments. We should consider moving forward this to merging.

File: kubernetes/chart/zulip/.gitignore
Comment (mrtndwrd) [reply kubernetes/chart/zulip/.gitignore]: No reason, I removed this from the gitignore and added Chart.lock

File: kubernetes/chart/zulip/values.yaml
Comment (mrtndwrd) [reply kubernetes/chart/zulip/values.yaml:126]: IMO this shouldn't block merging this PR. If you think it should, please suggest what affinity to put in and I'll do it. Or add a commit (IIRC I allowed the maintainer to push to this PR)

**Comment by mrtndwrd** [comment]
To keep up momentum I rebased this branch on `main` again. It's ready to merge ;)

File: kubernetes/chart/zulip/values.yaml
Comment (scratchmex) [reply kubernetes/chart/zulip/values.yaml:126]: I am happy to make a PR when this is merged

File: kubernetes/chart/zulip/README.md.gotmpl
Comment (alexmv) [inline kubernetes/chart/zulip/README.md.gotmpl:31]: The difficulty, at least with my play with minikube, is that because there's no SSL in the system, and the CSRF cookie is set as `Secure`, you can't get past the first form submission of the realm creation screen.

Did you successfully get this working in minikube?

File: kubernetes/chart/zulip/templates/NOTES.txt
Comment (alexmv) [inline kubernetes/chart/zulip/templates/NOTES.txt:14]: Can't we use the `$POD_NAME` we just defined?
```suggestion
  kubectl -n {{ .Release.Namespace }} exec -it "$POD_NAME" -c zulip -- sudo -u zulip /home/zulip/deployments/current/manage.py generate_realm_creation_link
```

File: kubernetes/chart/zulip/values.yaml
Comment (alexmv) [inline kubernetes/chart/zulip/values.yaml:147]: This comment seems cargo-culted in several places.  I don't believe this is a Zulip limitation in any way.  It's often a limitation of the _sending_ server these days, but nothing in Zulip requires this.

**Review by alexmv — COMMENTED** [review]
Thank you for all of your work on this -- and for your patience!

I left a few comments inline from installing minikube and kicking the tires a little, and reading through the YAML.

I think my biggest questions from the deployment point of view are the autoscaling bits -- I think we should omit them, unless you've got experience saying that they work.

The commit message looks like it needs to be updated a little.  When you talk about "first implementation," do you mean `kubernetes/manual/`, or something else?  If the former, it might be good to spell that out; if the latter, then folks (like me!) arriving at the PR now have no idea what the previous version was, so the comment doesn't help me understand it better. :)

We've also merged `SOCIAL_AUTH_OIDC_ENABLED_IDPS` as 6300fb4700b1c9305c3a4e5d9afcdb3d905ca746, so we can drop that from the commit message.

In general, the [Zulip commit message formatting](https://zulip.readthedocs.io/en/latest/contributing/version-control.html#commit-discipline) would be good to adhere to, though we're not generally as scrupulous about it in non-`zulip/zulip` repos.

File: kubernetes/chart/zulip/Chart.yaml
Comment (mrtndwrd) [reply kubernetes/chart/zulip/Chart.yaml:36]: Yes, I switched to zulip/zulip-postgresql:14 and I added a comment here.

File: kubernetes/chart/zulip/README.md.gotmpl
Comment (mrtndwrd) [reply kubernetes/chart/zulip/README.md.gotmpl:31]: I don't think I tested more than just getting the access screen and getting all pods ready. I'm using the chart in production, though, with prod. certificates.

If there's a way I can let Zulip ignore self signed certificates or http:// traffic I can apply that to make it work in Minikube, but I don't know a lot about Zulip internals.

File: kubernetes/chart/zulip/Chart.yaml
Comment (mrtndwrd) [reply kubernetes/chart/zulip/Chart.yaml]: updated to 6.0.16

File: kubernetes/chart/zulip/Chart.yaml
Comment (mrtndwrd) [reply kubernetes/chart/zulip/Chart.yaml]: updated to 16.8.7

File: kubernetes/chart/zulip/values.yaml
Comment (mrtndwrd) [reply kubernetes/chart/zulip/values.yaml]: Dropped.

File: kubernetes/chart/zulip/templates/hpa.yaml
Comment (mrtndwrd) [reply kubernetes/chart/zulip/templates/hpa.yaml]: dropped.

File: kubernetes/chart/zulip/values.yaml
Comment (mrtndwrd) [reply kubernetes/chart/zulip/values.yaml]: I think you're right. I removed the readinessProbe.

File: kubernetes/chart/zulip/Chart.yaml
Comment (mrtndwrd) [reply kubernetes/chart/zulip/Chart.yaml]: Updated to 8.32.0

File: kubernetes/chart/zulip/templates/NOTES.txt
Comment (mrtndwrd) [reply kubernetes/chart/zulip/templates/NOTES.txt]: Well, yes and no. I moved POD_NAME outside of the `if` so we can, although the people using this would have to copy-paste the commands into the terminal for them to work. I think it's clear enough now, though.

File: kubernetes/chart/zulip/values.yaml
Comment (mrtndwrd) [reply kubernetes/chart/zulip/values.yaml]: I think this came from #287. I removed it.

File: kubernetes/chart/zulip/README.md.gotmpl
Comment (mrtndwrd) [reply kubernetes/chart/zulip/README.md.gotmpl]: OK, I added instructions on how to set up cert-manager inside minikube and use the ingress addon to make zulip available on `zulip.local`, with a self-signed certificate. It worked perfectly in my tests, so I hope you'll get success too. If not, please reach out!

AFAIK that should absolve the last remaining blocker to getting this merged :) I'm pretty proud of the chart so far, and I'm really happy with the improvements based on your feedback. Thanks a lot :)

**Comment by alexmv** [comment]
Merged as 26c0b812e8f5ed1d1c5f3db16ce4aa2a303e5c9f -- thank you, @mrtndwrd (and everyone else!) for helping drive this forward.

I adjusted 5.1 to 5.2 before merging.

**Pull request closed without merging**
